### PR TITLE
samples: add exactly once subscribe sample

### PR DESCRIPTION
**Pull request by mmicatka**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/java-pubsub/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> ☕️

### CODE REVIEW

File: samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package pubsub;
+
+// [START pubsub_subscriber_exactly_once]
+// [START pubsub_create_subscription_with_exactly_once_delivery]
```
Comment (anguillanneuf) [inline samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java:20]: Please remove `pubsub_create_subscription_with_exactly_once_delivery`.

File: samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+                        default:
+                            break;
+                    }
+                };
+
+        Subscriber subscriber = null;
+        try {
+            subscriber = Subscriber.newBuilder(subscriptionName, receiverWithResponse)
+                    .setExactlyOnceDeliveryEnabled(true)
+                    .build();
```
Comment (anguillanneuf) [inline samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java:72]: We should allow setting exactly once delivery on the subscription. Is `.setExactlyOnceDeliveryEnabled()` updating the subscription under the hood to enable exactly once? If so, we need a comment here.

**Review by anguillanneuf — CHANGES_REQUESTED** [review]
- Please add a sample test. 
- `pubsub_create_subscription_with_exactly_once_delivery` is a different sample planned using `SubscriptionAdminClient`.

File: samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java (hunk 1 above)
Comment (mmicatka) [reply samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java]: Done

File: samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java (hunk 2 above)
Comment (mmicatka) [reply samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java]: Discussed offline. Resolving

File: samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+    // Instantiate an asynchronous message receiver.
+    MessageReceiverWithAckResponse receiverWithResponse =
+        (PubsubMessage message, AckReplyConsumerWithResponse consumerWithResponse) -> {
+          // Handle incoming message, then ack the received message, and receive the response
+
+          Future<AckResponse> ackResponseFuture = consumerWithResponse.ack();
+
+          // Retreive the completed future
+          AckResponse ackResponse = ackResponseFuture.get();
```
Comment (anguillanneuf) [inline samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java:56]: How should failures be handled here? What kind of exceptions will be thrown?

File: samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        ProjectSubscriptionName.of(projectId, subscriptionId);
+
+    // Instantiate an asynchronous message receiver.
+    MessageReceiverWithAckResponse receiverWithResponse =
+        (PubsubMessage message, AckReplyConsumerWithResponse consumerWithResponse) -> {
+          // Handle incoming message, then ack the received message, and receive the response
+
+          Future<AckResponse> ackResponseFuture = consumerWithResponse.ack();
+
+          // Retreive the completed future
```
Comment (anguillanneuf) [inline samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java:55]: Typo: "Retrieve"

File: samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  }
+
+  public static void subscribeWithExactlyOnceConsumerWithResponseExample(
+      String projectId, String topicId, String subscriptionId) {
+    // For subscriptions with exactly once enabled, the AckResponse will:
+    // return confirmed SUCCESS OR
+    // permanent failures
+    // For subscriptions without exactly once enabled the AckResponse will:
+    // return SUCCESS for messages sent OR
+    // permanent failures
```
Comment (anguillanneuf) [inline samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java:44]: Why is "SUCCESS" capitalized?  What's the difference between "confirmed SUCCESS" and "SUCCESS for message sent"?

**Review by anguillanneuf — CHANGES_REQUESTED** [review]

File: samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java (hunk 4 above)
Comment (mmicatka) [reply samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java]: Fixed

File: samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java (hunk 3 above)
Comment (mmicatka) [reply samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java]: Added a try-catch for the specific exceptions.

File: samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    // Instantiate an asynchronous message receiver.
+    MessageReceiverWithAckResponse receiverWithResponse =
+        (PubsubMessage message, AckReplyConsumerWithResponse consumerWithResponse) -> {
+          // Handle incoming message, then ack the received message, and receive the response
+          Future<AckResponse> ackResponseFuture = consumerWithResponse.ack();
+
+          try {
+            // Retrieve the completed future
+            AckResponse ackResponse = ackResponseFuture.get();
+            System.out.println("Id: " + message.getMessageId() + " " + ackResponse);
```
Comment (maheshgattani) [inline samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java:55]: How do we handle different ack responses? success/failures? Can we add a switch statement with logs?

File: samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+          // Handle incoming message, then ack the received message, and receive the response
+          Future<AckResponse> ackResponseFuture = consumerWithResponse.ack();
+
+          try {
+            // Retrieve the completed future
+            AckResponse ackResponse = ackResponseFuture.get();
+
+            switch (ackResponse) {
+              case SUCCESSFUL:
+                System.out.println("Successful MessageId: " + message.getMessageId());
```
Comment (maheshgattani) [inline samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java:58]: Message successfully acked: ....

Please update the language in the logs below as well.

File: samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+            // Something went wrong retrieving the future
+            System.out.println("Failed to retrieve future for Id: " + message.getMessageId());
+          }
+        };
+
+    Subscriber subscriber = null;
+    try {
+      subscriber =
+          Subscriber.newBuilder(subscriptionName, receiverWithResponse)
+              .setMinDurationPerAckExtension(Duration.ofSeconds(60))
```
Comment (maheshgattani) [inline samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java:85]: On second thought, let's remove this. Sorry for the churn.

File: samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java (hunk 6 above)
Comment (mmicatka) [reply samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java]: Added a switch statement + logs

File: samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java (hunk 7 above)
Comment (mmicatka) [reply samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java]: Updated

File: samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java (hunk 5 above)
Comment (mmicatka) [reply samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java]: I changed the comment wording

**Review by maheshgattani — APPROVED** [review]

File: samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+              case OTHER:
+                System.out.println(
+                    "Message failed to ack with a response of Other. Id: "
+                        + message.getMessageId());
+                break;
+              default:
+                break;
+            }
+          } catch (InterruptedException | ExecutionException e) {
+            // Something went wrong retrieving the future
```
Comment (anguillanneuf) [inline samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java:85]:
```suggestion
            // Something went wrong when retrieving the future.
```

File: samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+    // return success OR permanent failures
+    // For subscriptions without exactly once enabled the AckResponse will:
+    // return success for messages ack/nack'd OR permanent failures
+    ProjectSubscriptionName subscriptionName =
+        ProjectSubscriptionName.of(projectId, subscriptionId);
+
+    // Instantiate an asynchronous message receiver.
+    MessageReceiverWithAckResponse receiverWithResponse =
+        (PubsubMessage message, AckReplyConsumerWithResponse consumerWithResponse) -> {
+          // Handle incoming message, then ack the received message, and receive the response
```
Comment (anguillanneuf) [inline samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java:49]:
```suggestion
          // Handle incoming message, then ack the received message, and receive the response.
```

**Review by anguillanneuf — CHANGES_REQUESTED** [review]

File: samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+    subscribeWithExactlyOnceConsumerWithResponseExample(projectId, topicId, subscriptionId);
+  }
+
+  public static void subscribeWithExactlyOnceConsumerWithResponseExample(
+      String projectId, String topicId, String subscriptionId) {
+    // For subscriptions with exactly once enabled, the AckResponse will:
+    // return success OR permanent failures
+    // For subscriptions without exactly once enabled the AckResponse will:
+    // return success for messages ack/nack'd OR permanent failures
```
Comment (anguillanneuf) [inline samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java:42]: Can we move this to L53?

Comment (mmicatka) [reply samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java]: Moved.

File: samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+    ProjectSubscriptionName subscriptionName =
+        ProjectSubscriptionName.of(projectId, subscriptionId);
+
+    // Instantiate an asynchronous message receiver.
+    MessageReceiverWithAckResponse receiverWithResponse =
+        (PubsubMessage message, AckReplyConsumerWithResponse consumerWithResponse) -> {
```
[the remark this answers is not in the document]
Comment (mmicatka) [reply samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java:44]: I added a comment addressing this.

File: samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+          Future<AckResponse> ackResponseFuture = consumerWithResponse.ack();
+
+          try {
+            // Retrieve the completed future.
+            AckResponse ackResponse = ackResponseFuture.get();
+            
+            // For subscriptions with exactly once enabled, the AckResponse will
+            // return success OR permanent failures.
+            // For subscriptions without exactly once enabled the AckResponse will
+            // return success for messages ack/nack'd OR permanent failures.
```
Comment (anguillanneuf) [inline samples/snippets/src/main/java/pubsub/SubscribeWithExactlyOnceConsumerWithResponseExample.java:58]:
```suggestion
```

The wording somehow gives me the impression that in the first case of exactly once enabled, the AckResponse will return success for messages ack'ed OR permanent failures.  I'm not sure what's there to gain with this comment.

**Review by anguillanneuf — CHANGES_REQUESTED** [review]

**Review by anguillanneuf — APPROVED** [review]

**Comment by anguillanneuf** [comment]
We'll continue to see test failure until a new release of `libraries-bom`, to include `google-cloud-bom`, which is to include `google-cloud-pubsub` 1.116.0: https://repo1.maven.org/maven2/com/google/cloud/google-cloud-pubsub/1.116.0/

google-cloud-bom current is 0.169.0: https://repo1.maven.org/maven2/com/google/cloud/google-cloud-bom/
libraries-bom current is 24.4.0: https://repo1.maven.org/maven2/com/google/cloud/libraries-bom/

**Comment by anguillanneuf** [comment]
[google-cloud-bom](https://github.com/googleapis/java-cloud-bom) current is 0.170.0 now: https://repo1.maven.org/maven2/com/google/cloud/google-cloud-bom/
But waiting for the next release for [libraries-bom](https://github.com/GoogleCloudPlatform/cloud-opensource-java): https://repo1.maven.org/maven2/com/google/cloud/libraries-bom/
Which is used by samples: https://github.com/googleapis/java-pubsub/blob/5b3227f4013a5121a3db34702f90be5a9ba41748/samples/snippets/pom.xml#L46-L48

**Comment by anguillanneuf** [comment]
Sample test in `samples/snippets` continues to fail after library version bump to 1.116.0

https://github.com/googleapis/java-pubsub/blob/0b7d19fe39326c36152267e4d2a13a2b51530ac2/samples/snippets/pom.xml#L43-L59

**Pull request merged**